proto_proxy.c

```diff
@@ -19,17 +19,28 @@ static void *mcp_profile_alloc(void *ud, void *ptr, size_t osize, size_t nsize);
 /******** EXTERNAL FUNCTIONS ******/
 // functions starting with _ are breakouts for the public functions.
 
-// We want to ensure we slowly iterate through VM memory in two cases:
-// - If using the newer zero-allocation API, old backends will never close
-// after reload.
-// - If the server is completely idle and user is issuing reloads, they will
-// be confused when old backends do not close.
-// So every few seconds we push one minimal GC step.
-static void proxy_gc_poke(evutil_socket_t fd, short event, void *arg) {
-    LIBEVENT_THREAD *t = arg;
+// Attempting to push the lua GC algorithm to run outside of when requests are
+// being processed.
+void proxy_gc_poke(LIBEVENT_THREAD *t) {
     lua_State *L = t->L;
+    int vm_kb = lua_gc(L, LUA_GCCOUNT) + t->proxy_vm_extra_kb;
+    // only add GC debt with what we grew past the last reload size.
+    if (vm_kb > t->proxy_vm_last_kb * 1.5) {
+        // reset our counter once a full GC has completed
+        if (lua_gc(L, LUA_GCSTEP, vm_kb - t->proxy_vm_last_kb)) {
```

Comment:
this should step the amount of change since the last time this visited. currently seems to just chuck the full delta of debt in each time.

could also try:
- start when VM size doubles
- run minimal steps
- if VM size is triple, start increasing amount passed to LUA_GCSTEP

Reply:
also instrument this with GCISRUNNING to see if I'm accidentally unpausing it.